src/ctelnet.cpp

```diff
@@ -2817,27 +2820,44 @@ void cTelnet::processSocketData(char* in_buffer, int amount)
                                     datalen = 0;
                                     i = -1; // end the loop, this will make i and datalen the same.
                                 }
-                                //bugfix: BenH
+                                // compressed data starts in clean state
                                 iac = false;
                                 insb = false;
                                 command = "";
-                                ////////////////
                                 goto MAIN_LOOP_END;
                             }
                         }
                     }
                     //7. inside IAC SB
 
                     command += ch;
-                    if (iac && (ch == TN_SE)) //IAC SE - end of subcommand
-                    {
+                    if (iac && (ch == TN_SE)) { //IAC SE - end of subcommand
                         processTelnetCommand(command);
                         command = "";
                         iac = false;
                         insb = false;
-                    }
-                    if (iac) {
+                    } else if (iac && (ch == TN_IAC)) { // escaped TN_IAC
+                        command.pop_back();
                         iac = false;
+                    } else if (iac) {
+                        // Telnet options within a subcommand are not supported.
+                        // We assume that the SE went missing, possibly due to a
+                        // server bug, and try to recover.
+                        // Cf. https://github.com/Mudlet/Mudlet/issues/4385
+                        command.pop_back();
+                        command += TN_SE;
+                        processTelnetCommand(command);
+                        if (!mIncompleteSB) {
+                            mIncompleteSB = true;
+                            qWarning(R"("TELNET: the server did not properly terminate a subnegotiation (code %02x).\nSome data loss is likely. You should ask them to fix this.)",command[2]);
```

Comment:
```suggestion
                            qWarning(R"("TELNET: the server did not properly complete a subnegotiation (code %02x) and started another one instead.\nSome data loss is likely - mention this problem to the game admins.)", command[2]);
```

Reply:
Changed, except that I removed the "started another one instead" part because the trigger may be some other non-subnegotiation Telnet message.

Reply:
Fair - do you think you can put that in writing? "did not properly complete a subnegotiation" is straight to the point but does not give much context, I don't think a game admin reading this will have the idea of what happened and will just be confused.